ksql-engine/src/main/java/io/confluent/ksql/schema/inference/DefaultTopicInjector.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License; you may not use this file
+ * except in compliance with the License.  You may obtain a copy of the License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.schema.inference;
+
+import io.confluent.ksql.KsqlExecutionContext;
+import io.confluent.ksql.analyzer.Analysis;
+import io.confluent.ksql.analyzer.QueryAnalyzer;
+import io.confluent.ksql.ddl.DdlConfig;
+import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.metastore.StructuredDataSource;
+import io.confluent.ksql.parser.KsqlParser.PreparedStatement;
+import io.confluent.ksql.parser.SqlFormatter;
+import io.confluent.ksql.parser.tree.CreateAsSelect;
+import io.confluent.ksql.parser.tree.Expression;
+import io.confluent.ksql.parser.tree.IntegerLiteral;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.parser.tree.StringLiteral;
+import io.confluent.ksql.services.KafkaTopicClient;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlConstants;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.commons.lang3.ObjectUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.kafka.clients.admin.TopicDescription;
+
+public class DefaultTopicInjector implements TopicInjector {
+
+  private final KafkaTopicClient topicClient;
+  private final MetaStore metaStore;
+
+  public DefaultTopicInjector(
+      final KsqlExecutionContext executionContext
+  ) {
+    this(executionContext.getServiceContext().getTopicClient(), executionContext.getMetaStore());
+  }
+
+  DefaultTopicInjector(
+      final KafkaTopicClient topicClient,
+      final MetaStore metaStore) {
+    this.topicClient = topicClient;
+    this.metaStore = metaStore;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T extends Statement> PreparedStatement<T> forStatement(
+      final PreparedStatement<T> statement,
+      final KsqlConfig ksqlConfig,
+      final Map<String, Object> propertyOverrides) {
+    if (!(statement.getStatement() instanceof CreateAsSelect)) {
+      return statement;
+    }
+
+    final CreateAsSelect cas = (CreateAsSelect) statement.getStatement();
+
+    final String topic = topicName(cas, ksqlConfig);
+
+    final TopicDescription description = describeSource(cas, metaStore, ksqlConfig, topicClient);
```

Comment:
Can we pull the topic name out of the analysis as well?

Reply:
That's a good idea, but it seems inconsistent. In my test I have 
```sql
CREATE STREAM s1 WITH(kafka_topic='topic')AS SELECT * FROM SOURCE;
```

The `AstBuilder` has the following code:
```
  @Override
  public Node visitQuerySpecification(final SqlBaseParser.QuerySpecificationContext context) {
    final Table into;
    if (context.into != null) {
      into = (Table) visit(context.into);
    } else {
      // TODO: Generate a unique name
      final String intoName = "KSQL_Stream_" + System.currentTimeMillis();
      into = new Table(QualifiedName.of(intoName), true);
    }
```
Unless the query has a `INTO` statement, then this results in the unique KStream name and sets the second boolean in the `Table` constructor to `true`. This boolean represents `isStdOut`, and if that value is set then the `intoKafkaTopicName` will be `null`. In the example query in my unit test, it hits the `else` clause and I can't get the name from the analysis.

This might be a bug or just a poorly named method, but it seems to be more complicated than I had hoped. Thoughts?

Reply:
Talked offline - this was changed after #2529. I may be able to use it.

Reply:
Looked through the code again, we only set the `into` if the property is there:
```
    if (sink.getProperties().get(DdlConfig.KAFKA_TOPIC_NAME_PROPERTY) != null) {
      setIntoTopicName(sink);
    }
```
Since I don't want to change this behavior for all analysis, I will leave it as is.